src/Prometheus/Storage/Redis.php

```diff
@@ -348,6 +391,66 @@ private function collectHistograms(): array
         return $histograms;
     }
 
+    /**
+     * @return array
+     */
+    private function collectSummaries(): array
+    {
+        $keys = $this->redis->sMembers(self::$prefix . Summary::TYPE . self::PROMETHEUS_METRIC_KEYS_SUFFIX);
+        sort($keys);
+        $summaries = [];
+
+        foreach ($keys as $key) {
+            $raw = $this->redis->hGetAll($key);
+            $summary = json_decode($raw['__meta'], true);
+            unset($raw['__meta']);
+            $allLabelValues = [];
+
+            foreach (array_keys($raw) as $k) {
+                $d = json_decode($k, true);
+                $allLabelValues[] = $d['labelValues'];
+            }
+
+            // We need set semantics.
+            // This is the equivalent of array_unique but for arrays of arrays.
+            $allLabelValues = array_map('unserialize', array_unique(array_map('serialize', $allLabelValues)));
+            sort($allLabelValues);
+
+            foreach ($allLabelValues as $labelValues) {
+                $valuesKey = json_encode(['b' => 'values', 'labelValues' => $labelValues]);
+                $values = !empty($raw[$valuesKey]) ? explode(',', $raw[$valuesKey]) : [];
+                foreach ($summary['quantiles'] as $quantile) {
+                    $summary['samples'][] = [
+                        'name' => $summary['name'],
+                        'labelNames' => ['quantile'],
+                        'labelValues' => array_merge($labelValues, ['quantile' => $quantile]),
+                        'value' => Summary::getQuantile($quantile, $values)
+                    ];
+                }
+
+                // Add the count
+                $countKey = json_encode(['b' => 'count', 'labelValues' => $labelValues]);
+                $summary['samples'][] = [
+                    'name' => $summary['name'] . '_count',
+                    'labelNames' => [],
+                    'labelValues' => $labelValues,
+                    'value' => !empty($raw[$countKey]) ? (int)$raw[$countKey] : 0,
+                ];
+
+                // Add the sum
+                $sumKey = json_encode(['b' => 'sum', 'labelValues' => $labelValues]);
+                $summary['samples'][] = [
+                    'name' => $summary['name'] . '_sum',
+                    'labelNames' => [],
+                    'labelValues' => $labelValues,
+                    'value' => !empty($raw[$sumKey]) ? $raw[$sumKey] : 0,
+                ];
+            }
+            $summaries[] = $summary;
+        }
+        return $summaries;
+    }
```

Comment:
There are a LOT of nested loops in here, can we reduce these to improve performance?